src/classes/RD2_RecurringDonations_TDTM.cls

```diff
@@ -260,38 +260,103 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
             newRd.Day_of_Month__c != oldRd.Day_of_Month__c ||
             newRd.Status__c != oldRd.Status__c);
     }
+
     /***
-    * @description Handles setting next donation date for updated recurring donations
+    * @description Performs a full analysis and rebuild of the Schedule records for a Recurring Donation,
+    * but does not commit the schedule records to the database.  Uses the active schedules for the RD
+    * to calculate the Next Donation Date and store onto the RD record (in a Before Update context,
+    * thus avoiding the extra DML against the object).
     * @param rds Trigger.new records
     * @param oldRds Trigger.old records
     */
-    private void setNextDonationDateBeforeUpdate(List<npe03__Recurring_Donation__c> rds) {
+    private void setScheduleDependentFields(List<npe03__Recurring_Donation__c> rds,List<npe03__Recurring_Donation__c> oldRds) {
```

Comment:
What would you think about moving the setScheduleDependentFields, getExistingSchedules and addUpdatedSchedules methods into the ScheduleService class and just have a method in this class for `calculateNextDonationDateBeforeUpdate` that delegates to the service method? You could also consider slightly updating the existing `calculateNextDonationDates` method to support both insert & update. Either way is probably fine though, whatever you think makes the most sense.

Overall, I'm looking to keep the logic to handle the schedules and NDD in the service class and have the TDTM class delegate out as needed.

Reply:
I debated about these.  getExistingSchedules() would become the only method in the service that does dml.  Are you okay with that?  If so, I'm fine to relocate them.

Reply:
getExistingSchedules does a query, as opposed to DML, correct? Either way though I think it's ok.

Reply:
Right you are.  Query only.  Okay, I'll move these 3 methods to the service class.  Thanks!

Reply:
Looking at this a little more closely, I'm not sure I understand the logic of keeping synchronizeSchedulesAfterRDUpdate() in the trigger while moving setScheduleDependentFields().  Could you help me out with that?

Reply:
The `synchronizeSchedulesAfterRDUpdate` method delegates **all** processing to the service class and only collects sobject instances to add to the DMLWrapper. setScheduleDependentFields feels like it does more than that, especially with the other two methods. However, if you think the method is solid enough and doesn't having anything too business-logic specific we can keep it in this class. 

I was thinking of looking something like:
```
private void calculateNextDonationDateBeforeUpdate() {
    Map<Id, List<RecurringDonationSchedule__c>> schedulesByRdId = scheduleService.getUpdatedSchedules(rds, oldRds)
    for (npe03__Recurring_Donation__c rd : rds) {
        rd.npe03__Next_Payment_Date__c = scheduleService.getNextDonationDate(schedulesByRdId.get(rd.Id));
    }
```

Basically the service is class is called to calculate the schedule records and those schedule records are used to recalc the NDD. I'm not tied to the method names, but the concept is that logic to get the schedules is in the schedules service.

Reply:
Okay.  I buy that.  Integrating both existing schedules and new schedules is more like business logic.  I'm in.  I'll finish up and commit.

Reply:
Also moving getRecurringDonationsNeedingScheduleUpdates().  Don't want to call from schedule service to the trigger.